### PR TITLE
chore: use vitest for type testing

### LINKED ISSUE #2019: Use "assertType" from Vitest for type testing
### Scope

Adds a new behavior

### Compatibility

- [ ] This is a breaking change

## What

There's a more reliable way to write type testing that we have right now. Turns out, Vitest comes with first-class support for type testing. I suggest we migrate to the setup they recommend, utilizing utility types like `assertType` and `expectTypeOf`.

## Why

- We can assert on the actual types instead of relying on inference. 
- We will be able to assert on the actual expected errors instead of using `@ts-expect-error`.

## Materials

- https://vitest.dev/guide/testing-types.html#testing-types

### PR DESCRIPTION
**Pull request by KaiSpencer**
closes #2019

### CODE REVIEW

**Comment by KaiSpencer** [comment]
Hey @kettanaito,

I have made a first stab at this. Would it be possible for you to approve the workflow run so I can see if what I have done for the vitest config RE typescript versions is even remotely correct :D

File: package.json [hunk 1]
```diff
@@ -50,14 +50,14 @@
     "build": "pnpm clean && cross-env NODE_ENV=production tsup && pnpm patch:dts",
     "patch:dts": "node \"./config/scripts/patch-ts.js\"",
     "check:exports": "node \"./config/scripts/validate-esm.js\"",
-    "test": "pnpm test:unit && pnpm test:node && pnpm test:browser && pnpm test:native",
+    "test": "pnpm test:unit --run && pnpm test:node --run && pnpm test:browser && pnpm test:native --run",
```
Comment (kettanaito) [inline package.json]: Not sure this change belongs to this pull request. I'd rather we removed it and discussed it separately.

File: test/typings/vitest.config.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+const LIB_DIR = path.resolve(__dirname, '../../lib')
+
+export default defineConfig({
+  test: {
+    root: './test/typings',
+    include: ['**/*.test-d.ts'],
+    globals: true,
+    typecheck: {
+      checker: 'tsc',
+      tsconfig: (() => {
```
Comment (kettanaito) [inline test/typings/vitest.config.ts:16]: This is so neat! Love it.

File: test/typings/graphql.test-d.ts [hunk 3]
```diff
@@ -1,193 +1,211 @@
 import { parse } from 'graphql'
 import { graphql, HttpResponse, passthrough } from 'msw'
+import { describe, expectTypeOf, it } from 'vitest'
+
+describe('GraphQL handlers', () => {
```
Comment (kettanaito) [inline test/typings/graphql.test-d.ts]: Let's keep our type tests flat as we do out regular tests. Drop the `describe()` and make the test names reflect what they are testing. 

I don't mind mixing a bunch of behaviors in type tests. The price of maintaining them is rather low.

File: test/node/vitest.config.ts [hunk 4]
```diff
@@ -28,6 +28,6 @@ export default defineConfig({
      * If any tests attempt to run during that window,
      * they will fail, unable to resolve the "msw" import alias.
      */
-    singleThread: true,
+    poolOptions: { threads: { singleThread: true } },
```
Comment (kettanaito) [inline test/node/vitest.config.ts:31]: 👍 Good catch. I'm also considering moving away from this. We need this due to a single test packing the library. I want to introduce a new level of testing (that does the packing) and run it separately from the other tests. By doing so, we will finally be able to run Vitest in a watch mode, and also use multiple threads for the rest of the tests.

**Comment by kettanaito** [comment]
@KaiSpencer, looks like some of your changes were reflected in the `pnpm-lock.yaml`. Could you please run `pnpm install` and commit the changes to the lockfile? That should unblock the CI. Thanks.

File: package.json (hunk 1 above)
Comment (KaiSpencer) [reply package.json]: Sure thing, will remove

File: test/node/vitest.config.ts (hunk 4 above)
Comment (KaiSpencer) [reply test/node/vitest.config.ts:31]: That sounds great, yeah this was needed with the vitest upgrade

**Comment by KaiSpencer** [comment]
> @KaiSpencer, looks like some of your changes were reflected in the `pnpm-lock.yaml`. Could you please run `pnpm install` and commit the changes to the lockfile? That should unblock the CI. Thanks.

I have done this and committed the lockfile, so should be good to go again on CI now

**Comment by KaiSpencer** [comment]
Hey @kettanaito 
I am happy with the type testing now. but there are some browser tests failing
msw-api/setup-worker/life-cycle-events/on.test.ts

It seems the response.url is returning back as an empty string. Not quite sure why :D Any ideas?

**Comment by kettanaito** [comment]
@KaiSpencer, that's an unrelated issue (I think #2031).

**Review by kettanaito — APPROVED** [review]
This looks fantastic! Thank you so much, @KaiSpencer.

**Comment by KaiSpencer** [comment]
<3

**Comment by kettanaito** [comment]
You, sir, are incredible. Thank you for making this happen and welcome to contributors, @KaiSpencer 👏

**Comment by KaiSpencer** [comment]
> You, sir, are incredible. Thank you for making this happen and welcome to contributors, @KaiSpencer 👏

Thank you for the support throughout this, It was a nice Issue/PR to get my foot in the door 🙇‍♂️

**Comment by kettanaito** [comment]
If you see any other tasks that interest you, don't hesitate to leave a comment!

**Comment by KaiSpencer** [comment]
I'll take a look 👀

**Comment by kettanaito** [comment]
## Released: v2.2.10 🎉

This has been released in v2.2.10!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.2.10)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.2.10)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request merged**